### PR TITLE
Fix support for "Expand all quotes by default"

### PR DESCRIPTION
**Pull request by stanio**
Fix for #4.  Couldn't come up with more generic handling.  Note this changes the _"Expand all quotes by default"_ customization itself.

### CODE REVIEW

**Comment by stanio** [comment]
Here's a [test release](https://github.com/stanio/QuoteCollapse/releases/tag/v1.0.1-stanio): `quotecollapse-1.0.1-stanio.xpi` to try-out (@svitalsky).  Note the "Expand all quotes by default" rule:

```css
BODY.mailview blockquote[type="cite"][qctoggled="true"] {
```

needs to be changed to:

```css
BODY.mailview blockquote[type="cite"][qctoggled="false"] {
```

**Comment by svitalsky** [comment]
@stanio The problem seems to be fixed with this change, however it also “fixes” (I do not know whether intentionally or not—do not have time right now to look into the code—, hence the quote marks) the other thing I mentioned in the issue description, i.e. that in case of nested quotes collapsing inner quote also collapses outer one(s) by default. I didn't like this feature so I've changed—for myself—the default behaviour so that outer one gets collapsed only with Ctrl (Meta &c.) key. With your fix however that possibility—to collapse outer quote(s) together with the inner one—seems to be gone entirely: clicking on an expanded inner quote, whether with Ctrl (or other) key or without it—always collapses only that one, none other.

**Comment by stanio** [comment]
Thanks for the feedback.  I'm not sure I get it all correct so I'm asking for further assistance.  From the [extension home page](http://mjg.github.io/QuoteCollapse/index.html) I get:

-   Hold down SHIFT while clicking, and all quotes contained in the current one will be collapsed/expanded.
-   Hold down CTRL, and all quotes of the same nesting level will be toggled.
-   Hold down CTRL and SHIFT, and all quotes will be toggled.

As far as I've tried all of these appear to work as prescribed.  You seems to describe is you're trying to achieve the last one (_Hold down CTRL and SHIFT, and all quotes will be toggled_) just holding CTRL  – have this worked for you with v0.9 like that?

**Comment by svitalsky** [comment]
OK, sorry for the confusion: turns out that the “feature” I didn't like is actually part of the bug as well:
with **version 1.0 without customization** it does not happen (I've checked it just now), however **with customization** for expand-by-default it also causes (besides what was described in the issue) clicking on inner expanded quote to collapse all the outer (parent) quotes.

This can be prevented (in that configuration, i.e. when the customization is applied) by Ctrl key: with it click only collapses the quote being clicked.

I didn't test this particular quirk without customization before, therefore I thought until now that collapsing all the parents was a feature, standard behaviour.

So, in other words, it seems that your change fixes both parts of the bug: the one described in the issue as well as the one described here.

**Comment by svitalsky** [comment]
Oh, and an addendum for completeness: I haven't tested your change *without* customization at all, when I wrote it fixes the bug I only meant “when customization is applied”. ;-)

**Comment by stanio** [comment]
Version 1.0 introduced the feature of automatically expanding parent quotes when expanding a nested quote, given it is visible at all – with the introduction of _partially collapsed quotes_ this becomes more common use case.  This feature broken the _expand all by default_ customization, as it generally relies on the manipulated `qctoggled` attribute which this customization hacked around just reversing it visual representation.

This fix given here also happens to fix the original _expand all by default_ customization where you had to initially click a quote twice to make it collapse.  This is achieved by unifying the meaning of `qctoggled="true"`.

**Comment by stanio** [comment]
> I haven't tested your change without customization at all, when I wrote it fixes the bug I only meant “when customization is applied”

I'm generally testing few different scenarios:

-   Factory default
-   Partically collapse quotes
-   Expand all quotes by default

I may still miss other stuff.

As a side note to @mjg: I'll further revisit `isFullyExpanded(quote)` (in separate PR) to apply it on click (expand) of a quote so nested quotes may be automatically toggled as expanded if they happen to be fully revealed – this should make "partially collapsed" quotes with fully visible content immediately obvious.

**Comment by stanio** [comment]
> As a side note to @mjg: I'll further revisit isFullyExpanded(quote) (in separate PR)...

O.k.  Factored out a `_toggleFullyVisible(quote)` here, and there are the two bits I could include to this PR, or as a separate one:

-   stanio/QuoteCollapse@d691473bf5c0435dbc4ef6a5f5124119ef742b87
-   stanio/QuoteCollapse@791877d78dd5a52ff1114ed5b25738a07680ccc7

File: src/chrome/content/quotecollapse/quotecollapse.js [hunk 1]
```diff
@@ -72,16 +72,29 @@ blockquote[type="cite"][qctoggled="true"] {\n\
...
     messageDocument.getElementsByTagName("head").item(0).appendChild(StyleElement);
+
+    for(let quote of messageDocument.querySelectorAll("blockquote")) {
+      QuoteCollapse._toggleFullyVisible(quote);
+    }
+  },
+  
+  _toggleFullyVisible: function toggleFullyVisible(quote) {
+    if(quote.getAttribute("qctoggled") == "true")
+      return true;
```
Comment (stanio) [inline src/chrome/content/quotecollapse/quotecollapse.js]: I'll remove that `if` bit – causes somewhat unexpected behavior in certain situations for me.  It wasn't in the original patch, also.

**Comment by stanio** [comment]
Here's one more [test-drive package](/stanio/QuoteCollapse/releases/tag/v1.1-stanio) (`quotecollapse-1.1-stanio.xpi`) including current changes here, and the [two additional bits mentioned earlier](#issuecomment-386992835) – I'll include them to this PR on request.

**Comment by stanio** [comment]
All in all I want to say I'm very satisfied of the development around this fix as it addresses and improves on more cases including the _partially collapse quotes_.  I had previously tried to achieve something similar in much more complicated way to the point I've given up as too many unwanted side effects appeared, while this approach is much more straight-forward and simple.

**Comment by stanio** [comment]
O.k.  I've gone ahead and added the two additional bits mentioned earlier to this branch/PR:

-   Auto-update qctoggled="true" of nested quotes on expand
-   Auto-expand parent/nested on quote-level toggle

Michael, I can easily strip them back, or I could tweak them additionally (possible code style changes) if you find necessary.

---

So far suggested release notes could be:

-   Fixed/restored "Expand all quotes by default" support (broken in 1.0)
    Note though, the necessary style has tiny change in a backward-compatible manner – users need to adjust.
-   Improvements to the "Partially collapse quotes" uses

**Comment by mjg** [comment]
Thanks for all your work, @stanio , and sorry for the delays - I hardly find time for testing.
Changes look good to me, so I'm merging and releasing (and I see there's more in your pipeline).

**Comment by stanio** [comment]
Thanks for your support, Michael, and sorry for the burden caused by me.

> ... and I see there's more in your pipeline

I'll keep these private for now.  I have another extra (not committed yet) which might be interesting to integrate in near future, as it doesn't cause backward-compatibility problems as v1.0.

**Comment by stanio** [comment]
> -   Fixed/restored "Expand all quotes by default" support (broken in 1.0)
>     Note though, the necessary style has tiny change in a backward-compatible manner – users need to adjust.

_backward-compatible manner_ – I've really meant to say _backward-incompatible manner_
